File: core/ts/src/graphpf/manifold.ts
/**
 * ConstraintManifold — map constraint as a plug-in [Phase 2 / F2].
 *
 * RoadGraphManifold wraps shipping road-graph geometry (same scoring as mapProject).
 * CorridorManifold is a synthetic 1-D polyline + lateral tolerance demo only —
 * not field-validated for rail, undersea, or planetary navigation.
 */

import type { IRoadGraph, IGraphEdge } from "../types/index.ts";
import { haversineM, wrap360, llaToEnu, enuToLla, headingBetween, mulberry32, clamp } from "../math.ts";
import {
  nearestEdges,
  interpolateEdge,
  projectS,
  edgeById,
  outgoing,
} from "./geometry.ts";

export interface ManifoldState {
  lat: number;
  lon: number;
  alt: number;
  yaw_deg: number;
  edge_id: string;
  s: number;
  lateral_m: number;
}

export interface ConstraintManifold {
  project(state: ManifoldState): ManifoldState;
  neighbours(state: ManifoldState, distance_m: number): ManifoldState[];
  transition_cost(a: ManifoldState, b: ManifoldState): number;
  dim(): number;
}

function circDiffAbs(a: number, b: number): number {
  let d = wrap360(a) - wrap360(b);
  if (d > 180) d -= 360;
  if (d < -180) d += 360;
  return Math.abs(d);
}

function emptyState(): ManifoldState {
  return { lat: 0, lon: 0, alt: 0, yaw_deg: 0, edge_id: "", s: 0, lateral_m: 0 };
}

export class RoadGraphManifold implements ConstraintManifold {
  constructor(public readonly graph: IRoadGraph) {}

  dim(): number {
    return 1;
  }

  project(state: ManifoldState): ManifoldState {
    const out = { ...state };
    const cands = nearestEdges(this.graph, state.lat, state.lon, 5);
    if (!cands.length) return out;
    let best: IGraphEdge | null = null;
    let bestScore = 1e300;
    for (const e of cands) {
      const dH = circDiffAbs(e.heading_deg, state.yaw_deg);
      const p = interpolateEdge(this.graph, e, projectS(this.graph, e, state.lat, state.lon));
      const d = haversineM(state.lat, state.lon, p.lat, p.lon);
      const score = d + 0.15 * dH;
      if (score < bestScore) {
        bestScore = score;
        best = e;
      }
    }
    if (!best) return out;
    const s = projectS(this.graph, best, state.lat, state.lon);
    const p = interpolateEdge(this.graph, best, s);
    out.lat = p.lat;
    out.lon = p.lon;
    out.alt = p.alt;
    out.edge_id = best.id;
    out.s = s;
    out.yaw_deg = best.heading_deg;
    out.lateral_m = 0;
    return out;
  }

  neighbours(state: ManifoldState, distance_m: number): ManifoldState[] {
    const edge = edgeById(this.graph, state.edge_id);
    if (!edge) {
      return nearestEdges(this.graph, state.lat, state.lon, 8).map((e) => {
        const p = interpolateEdge(this.graph, e, 0);
        return {
          ...emptyState(),
          lat: p.lat,
          lon: p.lon,
          alt: p.alt,
          edge_id: e.id,
          yaw_deg: e.heading_deg,
          s: 0,
        };
      });
    }
    const remain = (1 - state.s) * edge.length_m;
    if (remain >= distance_m) {
      const n = { ...state };
      n.s = clamp(state.s + distance_m / Math.max(1e-3, edge.length_m), 0, 0.999);
      const p = interpolateEdge(this.graph, edge, n.s);
      n.lat = p.lat;
      n.lon = p.lon;
      n.alt = p.alt;
      n.yaw_deg = edge.heading_deg;
      return [n];
    }
    const outs = outgoing(this.graph, edge.to);
    if (!outs.length) {
      const n = { ...state, s: 0.999, yaw_deg: edge.heading_deg };
      const p = interpolateEdge(this.graph, edge, n.s);
      n.lat = p.lat;
      n.lon = p.lon;
      n.alt = p.alt;
      return [n];
    }
    return outs.map((o) => {
      const p = interpolateEdge(this.graph, o, 0);
      return {
        ...emptyState(),
        lat: p.lat,
        lon: p.lon,
        alt: p.alt,
        edge_id: o.id,
        yaw_deg: o.heading_deg,
        s: 0,
      };
    });
  }

  transition_cost(a: ManifoldState, b: ManifoldState): number {
    return haversineM(a.lat, a.lon, b.lat, b.lon) + 0.05 * circDiffAbs(a.yaw_deg, b.yaw_deg);
  }
}

interface Seg {
  a: { lat: number; lon: number; alt: number };
  b: { lat: number; lon: number; alt: number };
  len_m: number;
  cum_m: number;
  heading_deg: number;
}

/** SYNTHETIC DEMO ONLY — not field-validated. */
export class CorridorManifold implements ConstraintManifold {
  readonly lateral_tol_m: number;
  readonly polyline: { lat: number; lon: number; alt: number }[];
  readonly origin: { lat: number; lon: number; alt: number };
  private segs: Seg[] = [];
  private total_m = 0;

  constructor(
    polyline: { lat: number; lon: number; alt: number }[],
    lateral_tol_m: number,
    origin?: { lat: number; lon: number; alt: number },
  ) {
    this.polyline = polyline;
    this.lateral_tol_m = Math.max(0.1, lateral_tol_m);
    this.origin = origin ?? polyline[0] ?? { lat: 0, lon: 0, alt: 0 };
    let cum = 0;
    for (let i = 0; i + 1 < polyline.length; i++) {
      const a = polyline[i]!;
      const b = polyline[i + 1]!;
      let len = haversineM(a.lat, a.lon, b.lat, b.lon);
      if (len < 1e-6) len = 1e-6;
      cum += len;
      this.segs.push({
        a,
        b,
        len_m: len,
        cum_m: cum,
        heading_deg: headingBetween(a.lat, a.lon, b.lat, b.lon),
      });
    }
    this.total_m = cum;
  }

  dim(): number {
    return 1;
  }

  length_m(): number {
    return this.total_m;
  }

  pointAt(s_m: number): { lat: number; lon: number; alt: number } {
    if (!this.segs.length) return this.origin;
    s_m = clamp(s_m, 0, this.total_m);
    for (let i = 0; i < this.segs.length; i++) {
      const seg = this.segs[i]!;
      const start = seg.cum_m - seg.len_m;
      if (s_m <= seg.cum_m || i === this.segs.length - 1) {
        const u = (s_m - start) / seg.len_m;
        return {
          lat: seg.a.lat + (seg.b.lat - seg.a.lat) * u,
          lon: seg.a.lon + (seg.b.lon - seg.a.lon) * u,
          alt: seg.a.alt + (seg.b.alt - seg.a.alt) * u,
        };
      }
    }
    return this.polyline[this.polyline.length - 1]!;
  }

  project(state: ManifoldState): ManifoldState {
    const out = { ...state };
    if (!this.segs.length) return out;
    let bestD = Infinity;
    let bestSm = 0;
    let bestLat = 0;
    let bestH = 0;
    let bestSeg = 0;
    for (let i = 0; i < this.segs.length; i++) {
      const seg = this.segs[i]!;
      const pa = llaToEnu(this.origin, seg.a.lat, seg.a.lon, seg.a.alt);
      const pb = llaToEnu(this.origin, seg.b.lat, seg.b.lon, seg.b.alt);
      const pq = llaToEnu(this.origin, state.lat, state.lon, seg.a.alt);
      const vx = pb.e - pa.e;
      const vy = pb.n - pa.n;
      const wx = pq.e - pa.e;
      const wy = pq.n - pa.n;
      const den = vx * vx + vy * vy || 1;
      const t = clamp((wx * vx + wy * vy) / den, 0, 1);
      const cx = pa.e + t * vx;
      const cy = pa.n + t * vy;
      const latM = (wx * -vy + wy * vx) / Math.sqrt(den);
      const d = Math.hypot(pq.e - cx, pq.n - cy);
      if (d < bestD) {
        bestD = d;
        bestSm = seg.cum_m - seg.len_m + t * seg.len_m;
        bestLat = latM;
        bestH = seg.heading_deg;
        bestSeg = i;
      }
    }
    const clamped = clamp(bestLat, -this.lateral_tol_m, this.lateral_tol_m);
    const centre = this.pointAt(bestSm);
    const seg = this.segs[bestSeg]!;
    const pa = llaToEnu(this.origin, seg.a.lat, seg.a.lon, seg.a.alt);
    const pb = llaToEnu(this.origin, seg.b.lat, seg.b.lon, seg.b.alt);
    const vx = pb.e - pa.e;
    const vy = pb.n - pa.n;
    const len = Math.hypot(vx, vy) || 1;
    const nx = -vy / len;
    const ny = vx / len;
    const c = llaToEnu(this.origin, centre.lat, centre.lon, centre.alt);
    const snapped = enuToLla(this.origin, c.e + nx * clamped, c.n + ny * clamped, c.u);
    out.lat = snapped.lat;
    out.lon = snapped.lon;
    out.alt = snapped.alt;
    out.yaw_deg = bestH;
    out.s = this.total_m > 0 ? bestSm / this.total_m : 0;
    out.lateral_m = clamped;
    out.edge_id = String(bestSeg);
    return out;
  }

  neighbours(state: ManifoldState, distance_m: number): ManifoldState[] {
    if (!this.segs.length) return [];
    const s0 = clamp(state.s, 0, 1) * this.total_m;
    return [-distance_m, distance_m].map((ds) => {
      const p = this.pointAt(s0 + ds);
      return this.project({
        ...state,
        lat: p.lat,
        lon: p.lon,
        alt: p.alt,
        s: this.total_m > 0 ? clamp(s0 + ds, 0, this.total_m) / this.total_m : 0,
      });
    });
  }

  transition_cost(a: ManifoldState, b: ManifoldState): number {
    return Math.abs(a.s - b.s) * this.total_m + 0.5 * Math.abs(a.lateral_m - b.lateral_m);
  }
}

export interface ManifoldFilterConfig {
  n?: number;
  seed?: number;
}

/** Tiny manifold-only particle cloud for regression (not production GraphPF). */
export class ManifoldParticleFilter {
  private rng: () => number;
  private n: number;
  private particles: ManifoldState[] = [];
  private weights: number[] = [];

  constructor(
    private readonly m: ConstraintManifold,
    cfg: ManifoldFilterConfig = {},
  ) {
    this.n = cfg.n ?? 64;
    this.rng = mulberry32(cfg.seed ?? 26168);
  }

  seed(lat: number, lon: number, yawDeg: number): void {
    const base = this.m.project({ ...emptyState(), lat, lon, yaw_deg: yawDeg });
    const nbrs = this.m.neighbours(base, 5);
    this.particles = [];
    this.weights = [];
    for (let i = 0; i < this.n; i++) {
      let p = nbrs.length ? { ...nbrs[i % nbrs.length]! } : { ...base };
      p.s = clamp(p.s + 0.02 * (this.rng() - 0.5), 0, 1);
      this.particles.push(this.m.project(p));
      this.weights.push(1 / this.n);
    }
  }

  step(dt: number, speedMps: number, yawDeg: number): void {
    const dist = Math.max(0, speedMps) * dt;
    for (let i = 0; i < this.particles.length; i++) {
      const p = { ...this.particles[i]!, yaw_deg: yawDeg };
      const cands = this.m.neighbours(p, dist);
      if (!cands.length) {
        this.particles[i] = this.m.project(p);
        continue;
      }
      let best = cands[0]!;
      let bestCost = 1e300;
      for (const c of cands) {
        const cost = this.m.transition_cost(p, c) + 0.1 * circDiffAbs(c.yaw_deg, yawDeg);
        if (cost < bestCost) {
          bestCost = cost;
          best = c;
        }
      }
      this.weights[i]! *= Math.exp(-0.05 * bestCost);
      this.particles[i] = this.m.project(best);
    }
    let sum = this.weights.reduce((a, b) => a + b, 0);
    if (sum <= 0) sum = 1;
    this.weights = this.weights.map((w) => w / sum);
  }

  estimate(): ManifoldState {
    let best = 0;
    for (let i = 1; i < this.weights.length; i++) {
      if (this.weights[i]! > this.weights[best]!) best = i;
    }
    return this.particles[best] ?? emptyState();
  }

  lateralErrorM(lat: number, lon: number): number {
    const proj = this.m.project({ ...emptyState(), lat, lon });
    return haversineM(lat, lon, proj.lat, proj.lon);
  }
}

File: lab/nav/manifold.py
"""ConstraintManifold — map constraint as a plug-in [Phase 2 / F2].

``RoadGraphManifold`` wraps ``MapGraph`` projection / neighbour geometry used by
the shipping road-constrained particle filter. Numeric mapfilter outputs are
unchanged: ``RoadParticleFilter`` still calls ``MapGraph`` directly; this module
exposes the same operations behind a domain-general interface.

``CorridorManifold`` is a **synthetic demonstration only** (1-D polyline +
lateral tolerance). It is not field-validated for rail, undersea, or planetary
navigation.
"""

from __future__ import annotations

import math
from abc import ABC, abstractmethod
from dataclasses import dataclass, replace
from typing import Any, Sequence

import numpy as np

EARTH_R_M = 6_371_008.8


@dataclass
class ManifoldState:
    lat: float = 0.0
    lon: float = 0.0
    alt: float = 0.0
    yaw_deg: float = 0.0
    edge_id: str = ""
    s: float = 0.0
    lateral_m: float = 0.0
    # OSM edge index when wrapping MapGraph (avoids string round-trips).
    edge: int = -1


class ConstraintManifold(ABC):
    @abstractmethod
    def project(self, state: ManifoldState) -> ManifoldState: ...

    @abstractmethod
    def neighbours(self, state: ManifoldState, distance_m: float) -> list[ManifoldState]: ...

    @abstractmethod
    def transition_cost(self, a: ManifoldState, b: ManifoldState) -> float: ...

    @abstractmethod
    def dim(self) -> int: ...


def _haversine_m(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    dn = math.radians(lat2 - lat1) * EARTH_R_M
    de = math.radians(lon2 - lon1) * EARTH_R_M * math.cos(math.radians(lat1))
    return math.hypot(de, dn)


def _circ_diff_abs(a: float, b: float) -> float:
    d = (a - b + 180.0) % 360.0 - 180.0
    return abs(d)


class RoadGraphManifold(ConstraintManifold):
    """Existing OSM / MapGraph geometry behind the ConstraintManifold interface."""

    def __init__(self, graph: Any) -> None:
        self.g = graph

    def dim(self) -> int:
        return 1

    def project(self, state: ManifoldState) -> ManifoldState:
        out = ManifoldState(
            lat=state.lat,
            lon=state.lon,
            alt=state.alt,
            yaw_deg=state.yaw_deg,
            edge_id=state.edge_id,
            s=state.s,
            lateral_m=0.0,
            edge=state.edge,
        )
        edges = self.g.nearby_edges(state.lat, state.lon, 80.0)
        if edges.size == 0:
            return out
        best_score = float("inf")
        best: tuple[int, float, float, float, float, float] | None = None
        for e in edges:
            e = int(e)
            plat, plon, d, along, bearing = self.g.project(e, state.lat, state.lon)
            if not math.isfinite(d):
                continue
            dH = _circ_diff_abs(bearing, state.yaw_deg)
            score = d + 0.15 * dH
            if score < best_score:
                best_score = score
                best = (e, plat, plon, d, along, bearing)
        if best is None:
            return out
        e, plat, plon, _d, along, bearing = best
        total = float(self.g.edge_len_m[e])
        out.lat = plat
        out.lon = plon
        out.edge = e
        out.edge_id = str(e)
        out.s = along / max(total, 1e-9)
        out.yaw_deg = bearing
        out.lateral_m = 0.0
        return out

    def neighbours(self, state: ManifoldState, distance_m: float) -> list[ManifoldState]:
        e = state.edge if state.edge >= 0 else (int(state.edge_id) if state.edge_id.isdigit() else -1)
        if e < 0:
            out: list[ManifoldState] = []
            for ne in self.g.nearby_edges(state.lat, state.lon, max(distance_m, 40.0))[:8]:
                ne = int(ne)
                plat, plon, _, _, bearing = self.g.project(ne, state.lat, state.lon)
                out.append(
                    ManifoldState(
                        lat=plat,
                        lon=plon,
                        yaw_deg=bearing,
                        edge=ne,
                        edge_id=str(ne),
                        s=0.0,
                    )
                )
            return out
        total = float(self.g.edge_len_m[e])
        remain = (1.0 - state.s) * total
        if remain >= distance_m:
            s = min(state.s + distance_m / max(total, 1e-9), 0.999)
            # Re-project along edge by walking from u toward v.
            plat, plon, _, _, bearing = self.g.project(e, state.lat, state.lon)
            # Approximate: move toward edge end.
            u = int(self.g.edge_u[e])
            v = int(self.g.edge_v[e])
            lat = float(self.g.node_lat[u]) + s * (float(self.g.node_lat[v]) - float(self.g.node_lat[u]))
            lon = float(self.g.node_lon[u]) + s * (float(self.g.node_lon[v]) - float(self.g.node_lon[u]))
            return [
                ManifoldState(
                    lat=lat,
                    lon=lon,
                    yaw_deg=float(self.g.edge_bearing_deg[e]),
                    edge=e,
                    edge_id=str(e),
                    s=s,
                )
            ]
        # Junction: adjacent edges from v.
        node = int(self.g.edge_v[e])
        out = []
        for k in range(int(self.g.adj_ptr[node]), int(self.g.adj_ptr[node + 1])):
            code = int(self.g.adj_edge[k])
            fwd = code < self.g.n_edges
            ne = code if fwd else code - self.g.n_edges
            if ne == e:
                continue
            u = int(self.g.edge_u[ne]) if fwd else int(self.g.edge_v[ne])
            out.append(
                ManifoldState(
                    lat=float(self.g.node_lat[u]),
                    lon=float(self.g.node_lon[u]),
                    yaw_deg=float(self.g.edge_bearing_deg[ne]) if fwd else (float(self.g.edge_bearing_deg[ne]) + 180.0) % 360.0,
                    edge=ne,
                    edge_id=str(ne),
                    s=0.0,
                )
            )
        return out or [
            ManifoldState(
                lat=state.lat,
                lon=state.lon,
                yaw_deg=state.yaw_deg,
                edge=e,
                edge_id=str(e),
                s=0.999,
            )
        ]

    def transition_cost(self, a: ManifoldState, b: ManifoldState) -> float:
        return _haversine_m(a.lat, a.lon, b.lat, b.lon) + 0.05 * _circ_diff_abs(a.yaw_deg, b.yaw_deg)


class CorridorManifold(ConstraintManifold):
    """1-D polyline + hard lateral tolerance.

    SYNTHETIC DEMO ONLY — not field-validated for rail / channel / planetary use.
    """

    def __init__(
        self,
        polyline: Sequence[tuple[float, float]] | np.ndarray,
        lateral_tol_m: float,
        *,
        origin: tuple[float, float] | None = None,
    ) -> None:
        pts = [(float(p[0]), float(p[1])) for p in polyline]
        if len(pts) < 2:
            raise ValueError("CorridorManifold needs >= 2 polyline vertices")
        self.poly = pts
        self.lateral_tol_m = max(0.1, float(lateral_tol_m))
        self.origin = origin or pts[0]
        self._segs: list[tuple[tuple[float, float], tuple[float, float], float, float, float]] = []
        cum = 0.0
        for i in range(len(pts) - 1):
            a, b = pts[i], pts[i + 1]
            length = _haversine_m(a[0], a[1], b[0], b[1])
            if length < 1e-6:
                length = 1e-6
            cum += length
            de = math.radians(b[1] - a[1]) * EARTH_R_M * math.cos(math.radians(a[0]))
            dn = math.radians(b[0] - a[0]) * EARTH_R_M
            heading = math.degrees(math.atan2(de, dn)) % 360.0
            self._segs.append((a, b, length, cum, heading))
        self.total_m = cum

    def dim(self) -> int:
        return 1

    def length_m(self) -> float:
        return self.total_m

    def point_at(self, s_m: float) -> tuple[float, float]:
        s_m = min(max(s_m, 0.0), self.total_m)
        for i, (a, b, length, cum, _) in enumerate(self._segs):
            start = cum - length
            if s_m <= cum or i == len(self._segs) - 1:
                u = (s_m - start) / length
                return a[0] + (b[0] - a[0]) * u, a[1] + (b[1] - a[1]) * u
        return self.poly[-1]

    def project(self, state: ManifoldState) -> ManifoldState:
        best_d = float("inf")
        best_s_m = 0.0
        best_lat = 0.0
        best_h = 0.0
        best_i = 0
        olat, olon = self.origin
        for i, (a, b, length, cum, heading) in enumerate(self._segs):
            clat0 = math.cos(math.radians(olat))
            pae = (a[1] - olon) * EARTH_R_M * math.radians(1.0) * clat0
            pan = (a[0] - olat) * EARTH_R_M * math.radians(1.0)
            pbe = (b[1] - olon) * EARTH_R_M * math.radians(1.0) * clat0
            pbn = (b[0] - olat) * EARTH_R_M * math.radians(1.0)
            pqe = (state.lon - olon) * EARTH_R_M * math.radians(1.0) * clat0
            pqn = (state.lat - olat) * EARTH_R_M * math.radians(1.0)
            vx, vy = pbe - pae, pbn - pan
            wx, wy = pqe - pae, pqn - pan
            den = vx * vx + vy * vy
            t = (wx * vx + wy * vy) / (den if den > 0 else 1.0)
            t = 0.0 if t < 0.0 else (1.0 if t > 1.0 else t)
            cx, cy = pae + t * vx, pan + t * vy
            lat_m = (wx * (-vy) + wy * vx) / math.sqrt(den if den > 0 else 1.0)
            d = math.hypot(pqe - cx, pqn - cy)
            if d < best_d:
                best_d = d
                best_s_m = (cum - length) + t * length
                best_lat = lat_m
                best_h = heading
                best_i = i
        clamped = max(-self.lateral_tol_m, min(self.lateral_tol_m, best_lat))
        a, b, length, cum, _ = self._segs[best_i]
        clat0 = math.cos(math.radians(olat))
        pae = (a[1] - olon) * EARTH_R_M * math.radians(1.0) * clat0
        pan = (a[0] - olat) * EARTH_R_M * math.radians(1.0)
        pbe = (b[1] - olon) * EARTH_R_M * math.radians(1.0) * clat0
        pbn = (b[0] - olat) * EARTH_R_M * math.radians(1.0)
        vx, vy = pbe - pae, pbn - pan
        ln = math.hypot(vx, vy) or 1.0
        nx, ny = -vy / ln, vx / ln
        ce, cn = self._enu_of(*self.point_at(best_s_m))
        lat, lon = self._lla_of(ce + nx * clamped, cn + ny * clamped)
        return ManifoldState(
            lat=lat,
            lon=lon,
            yaw_deg=best_h,
            edge_id=str(best_i),
            edge=best_i,
            s=best_s_m / self.total_m if self.total_m > 0 else 0.0,
            lateral_m=clamped,
        )

    def _enu_of(self, lat: float, lon: float) -> tuple[float, float]:
        olat, olon = self.origin
        clat0 = math.cos(math.radians(olat))
        e = (lon - olon) * EARTH_R_M * math.radians(1.0) * clat0
        n = (lat - olat) * EARTH_R_M * math.radians(1.0)
        return e, n

    def _lla_of(self, e: float, n: float) -> tuple[float, float]:
        olat, olon = self.origin
        clat0 = math.cos(math.radians(olat))
        lat = olat + math.degrees(n / EARTH_R_M)
        lon = olon + math.degrees(e / (EARTH_R_M * max(clat0, 1e-9)))
        return lat, lon

    def neighbours(self, state: ManifoldState, distance_m: float) -> list[ManifoldState]:
        s0 = min(max(state.s, 0.0), 1.0) * self.total_m
        out: list[ManifoldState] = []
        for ds in (-distance_m, distance_m):
            lat, lon = self.point_at(s0 + ds)
            probe = ManifoldState(
                lat=lat,
                lon=lon,
                yaw_deg=state.yaw_deg,
                s=min(max(s0 + ds, 0.0), self.total_m) / self.total_m if self.total_m else 0.0,
            )
            out.append(self.project(probe))
        return out

    def transition_cost(self, a: ManifoldState, b: ManifoldState) -> float:
        return abs(a.s - b.s) * self.total_m + 0.5 * abs(a.lateral_m - b.lateral_m)


class ManifoldParticleFilter:
    """Tiny manifold-only particle cloud for regression (not RoadParticleFilter)."""

    def __init__(
        self,
        manifold: ConstraintManifold,
        *,
        n: int = 64,
        seed: int = 26168,
    ) -> None:
        self.m = manifold
        self.n = int(n)
        self.rng = np.random.default_rng(seed)
        self.particles: list[ManifoldState] = []
        self.weights: np.ndarray = np.array([])

    def seed(self, lat: float, lon: float, yaw_deg: float) -> None:
        base = self.m.project(ManifoldState(lat=lat, lon=lon, yaw_deg=yaw_deg))
        nbrs = self.m.neighbours(base, 5.0)
        self.particles = []
        for i in range(self.n):
            src = nbrs[i % len(nbrs)] if nbrs else base
            p = replace(src, s=min(max(src.s + 0.02 * (float(self.rng.random()) - 0.5), 0.0), 1.0))
            self.particles.append(self.m.project(p))
        self.weights = np.full(self.n, 1.0 / self.n)

    def step(self, dt: float, speed_mps: float, yaw_deg: float) -> None:
        dist = max(0.0, speed_mps) * dt
        for i, p in enumerate(self.particles):
            p.yaw_deg = yaw_deg
            cands = self.m.neighbours(p, dist)
            if not cands:
                self.particles[i] = self.m.project(p)
                continue
            best = cands[0]
            best_cost = float("inf")
            for c in cands:
                cost = self.m.transition_cost(p, c) + 0.1 * _circ_diff_abs(c.yaw_deg, yaw_deg)
                if cost < best_cost:
                    best_cost = cost
                    best = c
            self.weights[i] *= math.exp(-0.05 * best_cost)
            self.particles[i] = self.m.project(best)
        total = float(self.weights.sum())
        if total <= 0:
            total = 1.0
        self.weights /= total

    def estimate(self) -> ManifoldState:
        if not self.particles:
            return ManifoldState()
        return self.particles[int(np.argmax(self.weights))]

    def lateral_error_m(self, lat: float, lon: float) -> float:
        proj = self.m.project(ManifoldState(lat=lat, lon=lon))
        return _haversine_m(lat, lon, proj.lat, proj.lon)
